web/packages/teleterm/src/ui/components/OfflineGateway.tsx

```diff
@@ -16,41 +16,70 @@
 import Validation from 'shared/components/Validation';
 import { Attempt } from 'shared/hooks/useAsync';
 
-import { PortFieldInput } from './FieldInputs';
+import { useLogger } from 'teleterm/ui/hooks/useLogger';
 
 export function OfflineGateway(props: {
   connectAttempt: Attempt<void>;
-  /** Setting `isSupported` to false hides the port input. */
-  gatewayPort:
-    | { isSupported: true; defaultPort: string }
-    | { isSupported: false };
-  reconnect(port?: string): void;
+  reconnect(args: { localPort?: string }): void;
   /** Gateway target displayed in the UI, for example, 'cockroachdb'. */
   targetName: string;
   /** Gateway kind displayed in the UI, for example, 'database'. */
   gatewayKind: string;
-  portFieldLabel?: string;
+  /**
+   * Each callsite is expected to pass its own formSchema that parses form data from controls passed
+   * through renderFormControls. If the callsite doesn't pass any form data, it's expected to use
+   * emptyFormSchema. We cannot do params.formSchema || emptyFormSchema, as that would mess with
+   * type inference.
+   */
+  formSchema: z.ZodType<Partial<Record<FormFieldNames, string>>>;
+  /**
+   * renderFormControls allows each consumer to provide its own form fields with specific HTML form
+   * validation rules. The form fields are read through FormData – names on the inputs must match
+   * names available through the FormFields enum.
+   */
+  renderFormControls?: (isProcessing: boolean) => ReactNode;
 }) {
-  const portFieldLabel = props.portFieldLabel || 'Port (optional)';
-  const defaultPort = props.gatewayPort.isSupported
-    ? props.gatewayPort.defaultPort
-    : undefined;
+  const logger = useLogger('OfflineGateway');
+  const { reconnect } = props;
 
-  const [port, setPort] = useState(defaultPort);
   const [reconnectRequested, setReconnectRequested] = useState(false);
+  const [parseError, setParseError] = useState('');
 
   const isProcessing = props.connectAttempt.status === 'processing';
   const statusDescription = isProcessing ? 'being set up…' : 'offline.';
   const shouldShowReconnectControls =
     props.connectAttempt.status === 'error' || reconnectRequested;
 
+  const submitForm = (event: FormEvent<HTMLFormElement>) => {
+    event.preventDefault();
+    setReconnectRequested(true);
+    setParseError('');
+
+    const formData = new FormData(event.currentTarget);
+    const parseResult = props.formSchema.safeParse(
+      Object.fromEntries(formData.entries())
+    );
+
+    // Explicitly compare to false to make type inference work since strictNullChecks are off.
+    if (parseResult.success === false) {
+      // There's no need to show validation errors in the UI, since they come from a programmer
+      // error, not user inputting actually invalid data.
+      logger.error('Could not parse form', parseResult.error);
+      setParseError(`Could not submit form. See logs for more details.`);
+      return;
+    }
+
+    reconnect(parseResult.data);
```

Comment:
No pushback on the zod approach? I think it sucks!!! But I wasn't able to come up with anything better.

It'd likely be easier to use and think about if it was better integrated with our existing stuff (e.g. form components), but I didn't want to invest too much time into it upfront.

Reply:
> But now that I think about it, I'm wondering if we need that flexibility at all. The form data is passed to the `reconnect` function which has a defined type. This means that the callsites need to use identical field names. If so, why don't we render the controls inside the `OfflineGateway`?

Because different gateways store different data under `targetSubresourceName`. DB gateways have a db name there while app gateways have a target port. For now it of course doesn't matter, as `OfflineGateway` would show the field for `targetSubresourceName` only for app gateways. So technically we could get away with using just another prop for the target subresource name field.

Reply:
Since you already invested time in this, I would keep the current implementation, maybe it will make our lives easier in the future. I would only update the `reconnect` type.

Reply:
You mean the `reconnect` prop for `OfflineGateway`? I was going to say that I'm not sure if we should change it, as technically every `OfflineGateway` callsite is expected to pass the same `reconnect` function. But I just now realized that `DocumentGatewayKube` doesn't use `useGateway`.

Reply:
>You mean the reconnect prop for OfflineGateway?

Yes. Now it's type is `(args: { localPort?: string }): void` but we actually pass `targetSubresourceName` there as well. Technically, we can just add it there as an optional property, but maybe we could easily infer the whole type from `formSchema`?

Reply:
Looks good, thanks.